### PR TITLE
Topic/2024.1/range contracts refactoring

### PR DESCRIPTION
**Pull request by gfraiteur**

### CODE REVIEW

File: src/Metalama.Patterns.Contracts/InvariantAttribute.cs [hunk 1]
```diff
@@ -6,6 +6,21 @@
...
+/// <summary>
+/// Aspect that, when added to a parameterless <c>void</c> method, causes this method to be invoked after each property setter
+/// or method (but not property getters), except those annotated with <see cref="DoNotCheckInvariantsAttribute"/>. The target method
+/// must check any relevant condition and throw throw <see cref="InvariantViolationException"/> in case of violation.
+/// </summary>
+/// <remarks>
+/// <para>The aspect defines a protected method <c>VerifyInvariants</c> that calls all invariant methods defined in the object including
+/// in its base classes.</para>
+/// <para>
+///  If the <see cref="ContractOptions.IsInvariantSuspensionSupported"/> contract option is defined, the aspect also
```
Comment (svick) [inline src/Metalama.Patterns.Contracts/InvariantAttribute.cs:18]:
```suggestion
///  If the <see cref="ContractOptions.IsInvariantSuspensionSupported"/> contract option is defined, the aspect also introduces
```

File: src/Metalama.Patterns.Contracts/InvariantSuspensionCounter.cs [hunk 2]
```diff
@@ -1,20 +1,36 @@
...
 {
     private int _value;
 
+    /// <summary>
+    /// Decrements the counter. Note that this does not verify invariants when the counter is back to zero.
+    /// </summary>
     public void Dispose()
     {
         Interlocked.Decrement( ref this._value );
     }
```
Comment (svick) [inline src/Metalama.Patterns.Contracts/InvariantSuspensionCounter.cs:22]: Note that this implementation violates the `IDisposable` contract that:

> If an object's `Dispose` method is called more than once, the object must ignore all calls after the first one.

Though fixing that would probably require an allocation and I don't know if that's worth it.

File: src/Metalama.Patterns.Contracts/Numeric/DecimalBound.cs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    {
+        switch ( this._value )
+        {
+            case decimal.MinValue:
+                value = this._value;
+                conversionResult = ConversionResult.ExactlyMinValue;
+
+                return true;
+
+            case ulong.MaxValue:
```
Comment (svick) [inline src/Metalama.Patterns.Contracts/Numeric/DecimalBound.cs:322]:
```suggestion
            case decimal.MaxValue:
```

File: src/Metalama.Patterns.Contracts/Numeric/DoubleBound.cs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+                return false;
+
+            default:
+                value = (byte) this._value;
+
+                conversionResult = value switch
+                {
+                    byte.MinValue => ConversionResult.ExactlyMinValue,
+                    byte.MaxValue => ConversionResult.ExactlyMaxValue,
```
Comment (svick) [inline src/Metalama.Patterns.Contracts/Numeric/DoubleBound.cs:46]: This is going to say that `0.1` is `ExactlyMinValue`, but will only produce `ExactlyMaxValue` for exactly `255`. I don't think that was intended.

File: src/Metalama.Patterns.Contracts/Numeric/NumericBound.cs [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        this.IsAllowed = isAllowed;
+    }
+
+    public static NumericBound Create( long min, bool isAllowed = true ) => new Int64Bound( min, isAllowed );
+
+    public static NumericBound Create( ulong min, bool isAllowed = true ) => new UInt64Bound( min, isAllowed );
+
+    public static NumericBound Create( decimal min, bool isAllowed = true ) => new DecimalBound( min, isAllowed );
+
+    public static NumericBound Create( double min, bool isAllowed = true ) => new DoubleBound( min, isAllowed );
```
Comment (svick) [inline src/Metalama.Patterns.Contracts/Numeric/NumericBound.cs:28]: I think these parameters should be called `value`, not `min`.

**Comment by svick** [comment]
I think there should either be a test that checks the generated code for e.g. `[StrictlyLessThan(1)]`, or a run-time test for e.g. `[StrictlyLessThan(1)]` and `Math.BitDecrement(1)`.

**Review by svick — APPROVED** [review]

File: src/Metalama.Patterns.Contracts/Numeric/DoubleBound.cs (hunk 4 above)
Comment (gfraiteur) [reply src/Metalama.Patterns.Contracts/Numeric/DoubleBound.cs:46]: Yes this is by design.

File: src/Metalama.Patterns.Contracts/InvariantSuspensionCounter.cs (hunk 2 above)
Comment (gfraiteur) [reply src/Metalama.Patterns.Contracts/InvariantSuspensionCounter.cs:22]: Yeah saving allocation was my design goal. I have changed that to a disposable struct.

**Pull request merged**